finality-aleph/src/metrics/chain_state.rs

```diff
@@ -91,47 +133,105 @@ pub async fn run_chain_state_metrics<
         }
     };
 
+    let mut cache: LruCache<TP::TxHash, Instant> = LruCache::new(
+        NonZeroUsize::new(TRANSACTION_CACHE_SIZE).expect("the cache size is a non-zero constant"),
+    );
+
     let mut best_block_notifications = import_notifications
         .fuse()
         .filter(|notification| future::ready(notification.is_new_best));
     let mut finality_notifications = finality_notifications.fuse();
-
     let mut previous_best: Option<HE> = None;
+
     loop {
-        select! {
-            maybe_block = best_block_notifications.next() => {
-                match maybe_block {
-                    Some(block) => {
-                        let number = (*block.header.number()).saturated_into::<BlockNumber>();
-                        metrics.update_best_block(number);
-                        if let Some(reorg_len) = detect_reorgs(backend, previous_best, block.header.clone()) {
-                            metrics.report_reorg(reorg_len);
-                        }
-                        previous_best = Some(block.header);
+        iteration(
+            backend,
+            &mut best_block_notifications,
+            &mut finality_notifications,
+            &metrics,
+            &mut transaction_pool_info_provider,
+            &mut cache,
+            &mut previous_best,
+        )
+        .await;
+    }
+}
+
+async fn iteration<
+    X,
+    HE: HeaderT<Number = BlockNumber, Hash = B::Hash>,
+    B: BlockT<Header = HE, Extrinsic = X>,
+    BE: HeaderMetadata<B> + BlockBackend<B>,
+    TP: TransactionPoolInfoProvider<Extrinsic = X>,
+    BBN: Stream<Item = BlockImportNotification<B>> + Unpin,
+    FN: Stream<Item = FinalityNotification<B>> + Unpin,
+>(
+    backend: &BE,
+    best_block_notifications: &mut BBN,
+    finality_notifications: &mut FN,
+    metrics: &ChainStateMetrics,
+    transaction_pool_info_provider: &mut TP,
+    cache: &mut LruCache<TP::TxHash, Instant>,
+    previous_best: &mut Option<HE>,
+) {
+    select! {
+        maybe_block = best_block_notifications.next() => {
+            match maybe_block {
+                Some(block) => {
+                    let number = (*block.header.number()).saturated_into::<BlockNumber>();
+                    metrics.update_best_block(number);
+                    if let Some(reorg_len) = detect_reorgs(backend, previous_best.clone(), block.header.clone()) {
+                        metrics.report_reorg(reorg_len);
                     }
-                    None => {
-                        warn!(target: LOG_TARGET, "Import notification stream ended unexpectedly");
+                    if let Ok(Some(body)) = backend.block_body(block.hash) {
+                        report_transaction_included_in_block(transaction_pool_info_provider, &body, metrics, cache);
```

Comment:
I am not sure if this is the proper way to observe transactions. Notice you only consider transactions which are contained in **best blocks**. What seems more reasonable to me, and what you have indicated in the PR description is counting transaction in all **imported blocks**. It's easy to align to this - at the beginning of `run_chain_state_metrics` you collapse `import_notification` into `best_block_notifications` - don't do that, accept all import notifications through the stream, and let any logic for best block change reside under an `if notification.is_new_best` statement here in the `select!` branch.

In healthy conditions I guess that the set of **best blocks** should be almost equal to the set of **imported blocks**, but in case a long reorg you'll potentially miss a lot of transactions.

Reply:
Yeah, this is a very good point.

I would suggest avoiding too complex logic though and treat the SLO doc literally – so consider transactions included in **any** imported block (even on a fork) to have been included (I'm not sure if this is what you were proposing). I think this will require some changes to avoid duplicate inclusion reports – I think the easiest way is to have a second `LruCache` remembering which transactions have already been counted and filtering through this. As a nice side-effect this lets you get rid of the ugly `bool` argument to the reporting function, since it simultaneously tracks that.

Reply:
My opinion isn't strong tho, so if you think this creates more confusion than it brings value, I can change the meaning of _being in block_ to as you suggested.

Reply:
@timorleph I was too late and haven't seen your comment, but I assume you're opting for simplifying? :smile:

Reply:
I think the fact that we might count some transactions as included only when they get finalized is a tad of a problem – not only is it inconsistent with the SLO definition, but if it happens it's likely to artificially inflate the average time to inclusion (although we will likely be alerting about something else in such a case anyway). That being said the current solution is not _terrible_, so I'll not be insistent about changing it the way described above. So, change this if you want or wait for @woocash2 to weight in.

Reply:
I agree that counting transactions at `finalized` events is going to wrongly inflate the inclusion time. It takes some time to perform a reorg and this time would contribute to the inclusion time of each transaction on the target branch. Each reorg will cause spikes in transaction metrics, and we want the spikes to indicate a problem with block inclusion mechanism, not to find out that there was a reorg :stuck_out_tongue:.

Counting transactions on every import notification doesn't have this problem. Also, having in mind that in healthy scenario **best blocks** and **imported blocks** are almost the same (i.e. `notification.is_new_best` will almost always be true), most of the time _import-based_ approach will behave exactly as your newly proposed _best-block-based_ approach :stuck_out_tongue:.

Reply:
>  Also, having in mind that in healthy scenario best blocks and imported blocks are almost the same 

Forks are exactly the complement of such healthy scenarios, so we could argue similarly that reorgs almost never occur :stuck_out_tongue: 
Nevertheless, reporting transactions in any block is also acceptable for me and I changed accordingly.

Reply:
My argument that _import-based approach_ ~= _best-block-based approach_ was just to try to convince you that you'll get what you have proposed most of the time. But far more important to me is that in these unhealthy scenarios import-based aligns with SLO much better :stuck_out_tongue:.